From f911d5d2140b8022249921811853264f9cba57c3 Mon Sep 17 00:00:00 2001
From: Subbaraya Sundeep <sbhatta@marvell.com>
Date: Mon, 19 Aug 2019 18:13:15 +0530
Subject: [PATCH 322/767] octeontx2-af: Use nix_smq_flush function

commit f31c3d4bd2b2c467b38940f56a361ea2f4002270 from
git@git.assembla.com:cavium/WindRiver.linux.git

nix_smq_flush function is available to flush all
metadescriptors/packets from SMQ through PSE and
the send data path. Hence use it wherever
required.

Change-Id: I9b2c85971b95780475182e8a5bd8d82a7ee9dd42
Signed-off-by: Subbaraya Sundeep <sbhatta@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/14558
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 .../net/ethernet/marvell/octeontx2/af/rvu_nix.c  | 16 +++++++---------
 1 file changed, 7 insertions(+), 9 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu_nix.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu_nix.c
index c16373ff2ab4..262a8cafa4ab 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu_nix.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu_nix.c
@@ -1903,18 +1903,16 @@ int rvu_mbox_handler_nix_txschq_cfg(struct rvu *rvu,
 			mutex_unlock(&rvu->rsrc_lock);
 		}
 
-		rvu_write64(rvu, blkaddr, reg, regval);
-
-		/* Check for SMQ flush, if so, poll for its completion */
+		/* SMQ flush is special hence split register writes such
+		 * that flush first and write rest of the bits later.
+		 */
 		if (schq_regbase == NIX_AF_SMQX_CFG(0) &&
 		    (regval & BIT_ULL(49))) {
-			err = rvu_poll_reg(rvu, blkaddr,
-					   reg, BIT_ULL(49), true);
-			if (err) {
-				rvu_nix_txsch_unlock(nix_hw);
-				return NIX_AF_SMQ_FLUSH_FAILED;
-			}
+			schq = TXSCHQ_IDX(reg, TXSCHQ_IDX_SHIFT);
+			nix_smq_flush(rvu, blkaddr, schq, pcifunc, nixlf);
+			regval &= ~BIT_ULL(49);
 		}
+		rvu_write64(rvu, blkaddr, reg, regval);
 	}
 
 	rvu_nix_txsch_config_changed(nix_hw);
-- 
2.31.1

